Add nsTick() function to return elapsed nanoseconds

ghalliday opened the pull request:
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday committed:
Add nsTick() function to return elapsed nanoseconds

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-15598, label: issue26864 starts now.
The reason of this test is: Forced to re-test.
Commit ID: b5b0bf5779d08f67f3b9eba95e1d439b9710a862
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-31.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-31.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: b5b0bf5779d08f67f3b9eba95e1d439b9710a862
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.4.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |222 |222 |0 |0 |0 |63 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |4 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |17 sec  (00:00:17) |
|test (hthor) |983 |983 |0 |781 sec  (00:13:01) |
|test (thor) |893 |893 |0 |994 sec  (00:16:34) |
|test (roxie) |1052 |1052 |0 |873 sec  (00:14:33) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  501 sec (00:08:21) |  103 sec (00:01:43) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1307 sec (00:21:47) |   19 sec (00:00:19) | 1981 sec (00:33:01) |

mckellyln commented on system/jlib/jutil.cpp:
```diff
@@ -1647,8 +1651,26 @@ unsigned usTick()
+    if (!use_gettimeofday) {
+        timespec tm;
+        if (clock_gettime(CLOCK_MONOTONIC, &tm)>=0)
+            return ((unsigned __int64)tm.tv_sec)*1000000000+tm.tv_nsec;
```
Should we add a 'LL' to the 1000000000 constant ?  Probably not needed since the first term in cast to the proper size.

ghalliday replied:
I think this safe, but it would be a sensible change

mckellyln commented on system/jlib/jutil.cpp:
```diff
@@ -1665,15 +1687,21 @@ unsigned msTick()
+{
+    struct timeval tm;
+    gettimeofday(&tm,NULL);
+    return tm.tv_sec*1000000000+tm.tv_usec*1000;
```
same question here

ghalliday replied:
Got catch - I think this is a real problem.

richardkchapman replied:
Your comments suggest you had changed this in latest push, but it appears unchanged.

ghalliday replied:
Good catches.  Fixed.

mckellyln reviewed (commented):
Just the one question about the constant in the muliply

ghalliday commented:
@mckellyln I have changed to use U64C(), and expanded out the windows definition.  I also squashed since a small change, but it would probably have been better not to.

HPCCSmoketest commented:
Process of PR-15598, label: issue26864 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 09a9792eef808778ed3c10f50d2cedd79f76ce73
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-175.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-175.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: 09a9792eef808778ed3c10f50d2cedd79f76ce73
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.4.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |222 |222 |0 |0 |0 |62 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |4 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |983 |983 |0 |818 sec  (00:13:38) |
|test (thor) |893 |893 |0 |990 sec  (00:16:30) |
|test (roxie) |1052 |1052 |0 |870 sec  (00:14:30) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  498 sec (00:08:18) |  106 sec (00:01:46) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1302 sec (00:21:42) |   18 sec (00:00:18) | 1976 sec (00:32:56) |

richardkchapman commented on system/jlib/jutil.cpp:
```diff
@@ -1617,6 +1617,20 @@ unsigned usTick()
+    }
+    if (!QueryPerformanceCounter(&v))
+        return 0;
+    return (unsigned) ((v.QuadPart*U64C(1000000000))/freq); // This is unlikely to cleanly wrap
```
question: Why are you casting to unsigned when the return type is unsigned __int64?

HPCCSmoketest commented:
Process of PR-15598, label: issue26864 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 7fa543d46ae486dd36ca05002c6e17afe12faa25
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-113.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-113.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: 7fa543d46ae486dd36ca05002c6e17afe12faa25
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.4.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |222 |222 |0 |0 |0 |63 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |0 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |983 |983 |0 |815 sec  (00:13:35) |
|test (thor) |893 |893 |0 |986 sec  (00:16:26) |
|test (roxie) |1052 |1052 |0 |925 sec  (00:15:25) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  504 sec (00:08:24) |  108 sec (00:01:48) |   19 sec (00:00:19) |   16 sec (00:00:16) | 1298 sec (00:21:38) |   19 sec (00:00:19) | 1979 sec (00:32:59) |

mckellyln reviewed (approved):
Looks good.

richardkchapman commented on system/jlib/jutil.cpp:
```diff
@@ -1617,6 +1617,20 @@ unsigned usTick()
+    }
+    if (!QueryPerformanceCounter(&v))
+        return 0;
+    return ((v.QuadPart*U64C(1000000000))/freq); // This is unlikely to cleanly wrap
```
picky: There seem to be a spurious set of brackets here

HPCCSmoketest commented:
Process of PR-15598, label: issue26864 starts now.
The reason of this test is: Base branch changed.
Commit ID: d31127a08d0b0ef5553262ed497396d66e311ebd
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-19.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-19.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: d31127a08d0b0ef5553262ed497396d66e311ebd
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.4.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |222 |222 |0 |0 |0 |62 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |4 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |17 sec  (00:00:17) |
|test (hthor) |983 |983 |0 |814 sec  (00:13:34) |
|test (thor) |893 |893 |0 |933 sec  (00:15:33) |
|test (roxie) |1052 |1052 |0 |872 sec  (00:14:32) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  501 sec (00:08:21) |  103 sec (00:01:43) |   25 sec (00:00:25) |   16 sec (00:00:16) | 1244 sec (00:20:44) |   18 sec (00:00:18) | 1924 sec (00:32:04) |

HPCCSmoketest commented:
Process of PR-15598, label: issue26864 starts now.
The reason of this test is: Forced to re-test.
Commit ID: b162692d019d96d958b4779f6f0cc620cb45341c
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-84.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-84.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.18.0,  cUrl: 7.67.0,  node.js: v16.13.0,  npm: 8.1.0
Sha: b162692d019d96d958b4779f6f0cc620cb45341c
Containerized:False
Build: success
Milestone:Install hpccsystems-platform-community_8.4.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |222 |222 |0 |0 |0 |61 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |38 sec  (00:00:38) |
|setup (roxie) |19 |19 |0 |18 sec  (00:00:18) |
|test (hthor) |983 |983 |0 |791 sec  (00:13:11) |
|test (thor) |893 |893 |0 |943 sec  (00:15:43) |
|test (roxie) |1052 |1052 |0 |883 sec  (00:14:43) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  498 sec (00:08:18) |  102 sec (00:01:42) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1253 sec (00:20:53) |   18 sec (00:00:18) | 1923 sec (00:32:03) |
